### ISSUE TITLE
Use path of custom_target as string while creating a dependency

### ISSUE THREAD

**Issue by 0xabu** [mesonbuild/meson#10241]
**Describe the bug**
The documentation for [custom_tgt.full_path()](https://mesonbuild.com/Reference-manual_returned_custom_tgt.html#custom_tgtfull_path) states:
> In most cases using the object itself will do the same job as this and will also allow Meson to setup inter-target dependencies correctly. **Please file a bug if that doesn't work for you.**

This is that bug. I want to pass the output path of a custom target as a string to the C compiler (in a CPP macro definition). I can do so with full_path(), but I cannot figure out how to also get a dependency.

**To Reproduce**
`magic_blob.S` contains directives such as:
```
magic_bin_start:
    .incbin MAGIC_BIN_PATH
magic_bin_end:
```

My `meson.build` links an executable (magic_elf) more-or-less normally, then embeds it in another program, using a custom objcopy invocation in between. To build the second program, I just need the CPP macro MAGIC_BIN_PATH to be set to the build-time path of a file that was created by a custom_target. The following *almost* works:

```
magic_elf = executable(...)

objcopy = find_program('objcopy')

magic_bin = custom_target(
  'magic.bin',
  output: 'magic.bin',
  input: magic_elf,
  command: [objcopy, '-O', 'binary', '@INPUT@', '@OUTPUT@'],
)

executable(
  'myprog',
  files('myprog.cpp', 'magic_blob.S'),
  cpp_args: ['-DMAGIC_BIN_PATH="' + magic_bin.full_path() + '"'])
```

**Expected behavior**
The problem with this is that it doesn't emit a dependency from myprog to the custom target (so magic.bin won't be created in a clean build, for example). If I use `magic_bin` in place of `magic_bin.full_path()` I get the error:

```
../meson.build:19:0: ERROR: The `+` operator of str does not accept objects of type CustomTarget (<CustomTarget magic.bin@cus: ['/usr/bin/objcopy', '-O', 'binary',  '@INPUT@', '@OUTPUT@']>)
```

**system parameters**
* Native build
* Ubuntu 20.04.4
* Python 3.8.10
* Meson 0.62.0
* Ninja 1.10.0

**Comment by 0xabu** [comment]
@eli-schwartz thanks. Adding it to the list of input makes the clean build work, but incremental builds are broken as you noted. Is there any workaround to just make an explicit dependency? I looked at declare_dependency but it seems inappropriate.

**Comment by eli-schwartz** [comment]
No, it's currently impossible to add an arbitrary incremental rebuild dependency to a file that is part of the internal implementation of `executable()`. There is a PR to fix that issue via two-stage compile so we get useful depfiles, which is on my list of things to review...

A dirty, filthy, terrible hack that is backwards compatible to very old versions of Meson, would be to generate a pointless no-op `pointless-header.h` file in the same custom_target that makes magic.bin (or, I suppose, in a custom_target that depends on it).

Then use the C preprocessor to include it in your asm file, which does nothing other than force magic_blob.S to be considered out of date every time `pointless-header.h` is rebuilt.
